torchtune/modules/layer_dropout.py

```diff
@@ -0,0 +1,302 @@
+    prob_layer_scale: Optional[ScaleType] = ScaleType.UNIFORM,
+    layers_str: Optional[str] = None,
+    disable_on_eval: Optional[bool] = True,
+) -> None:
+    """
+    Prepare a model for layer dropout by wrapping each layer with a ModuleLayerDropoutWrapper.
+    This function takes in a model, the maximum probability of dropping a layer,
+    the scaling type for the layer dropout probability, a string specifying which
+    layers to apply dropout to, and a boolean indicating whether to disable dropout
+    during evaluation. It then wraps each layer of the model inplace with a
+    ModuleLayerDropoutWrapper, which applies layer dropout to the input tensor.
+    Args:
+        model (torch.nn.Module): The model to prepare for layer dropout.
+        prob_max (float): The maximum probability of dropping a layer. Defaults to 0.0.
+        prob_layer_scale (Optional[ScaleType]): The scaling type for the dropout probability
+            across layers. Defaults to ScaleType.UNIFORM.
+        layers_str (Optional[str]): A string specifying which layers to apply dropout to.
+            Defaults to None which means apply to all layers.
+        disable_on_eval (Optional[bool]): Whether to disable dropout during evaluation. Defaults to True.
+    Returns:
+        None
+    Example:
+        >>> import torch
+        >>> from torch import nn
+        >>> # Define a simple model
+        >>> class MyModel(nn.Module):
+        ...     def __init__(self):
+        ...         super().__init__()
+        ...         self.layers = nn.ModuleList([
+        ...             nn.Linear(5, 3),
+        ...             nn.Linear(3, 2),
+        ...             nn.Linear(2, 1),
+        ...             nn.Linear(1, 2),
+        ...             nn.Linear(2, 3),
+        ...         ])
+        ...
+        ...     def forward(self, x):
+        ...         for layer in self.layers:
+        ...             x = layer(x)
+        ...         return x
+        >>> model = MyModel()
+        >>> # Apply layer dropout uniformly to all layers
+        >>> prepare_layer_dropout(model, prob_max=0.2, prob_layer_scale=ScaleType.UNIFORM)
+        >>> # Apply layer dropout every other layer, as described in LayerDrop paper
+            (Fan et al., https://arxiv.org/abs/1909.11556v1)
+        >>> prepare_layer_dropout(model, prob_max=0.2, prob_layer_scale=ScaleType.UNIFORM, layers_str="::2")
+        >>> # Apply layer dropout that increases linearly across layers, as described in Progressive Layer
+            Dropout paper (Zhang et al., https://arxiv.org/abs/2010.13369)
+        >>> prepare_layer_dropout(model, prob_max=0.2, prob_layer_scale=ScaleType.LINEAR)
+        >>> # Apply layer dropout that increases exponentially across layers, as described in
+            LayerSkip paper (Elhoushi et al., https://arxiv.org/abs/2404.16710)
+        >>> prepare_layer_dropout(model, prob_max=0.2, prob_layer_scale=ScaleType.EXP)
+    """
+    num_layers = len(model.layers)
```

Comment:
nit but you're assuming that model has `layers` attributed defined. I don't see an obvious way around it though, maybe just add an explicit assert to verify this?

Reply:
Good point. In 78aff5af2cc79f74d977510dbfa703af97aab033 , I changed the function to take `layers` as input rather than `model`.
This way the function makes no assumptions, and we hand over both the flexibility and  responsibility to the user to provide a list of layers to the function.